File: src/etch/bindings/csharp/transport/TcpTransportFactory.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Net.Sockets;
using Etch.Support;
using Etch.Util;

namespace Etch.Transport
{
    public class TcpTransportFactory : TransportFactory
    {

        public override DefaultDeliveryService NewTransport( Etch.Util.URL uri, Resources resources )
        {
            Socket socket = ( Socket ) resources.Get( "SOCKET" );

            TcpConnection c = new TcpConnection( uri, resources, socket );

            Packetizer p = new Packetizer( uri, resources );
            c.SetHandler( p );

            Messagizer m = new Messagizer( uri, resources );
            p.SetHandler( m );

            MailboxManager r = new MailboxManager( uri, resources );
            m.SetHandler( r );

            DefaultDeliveryService d = new DefaultDeliveryService( uri, resources );
            r.SetHandler( d );

            return d;
        }

        public override Transport NewListener( URL uri, Resources resources, ServerFactory factory )
        {
            MyListenerHandler lh = new MyListenerHandler( this, uri, resources, factory );
            Listener l = new Listener( uri );
            l.SetHandler( lh );
            return l;
        }

        public class MyListenerHandler : ListenerHandler
        {
            private TcpTransportFactory _ttf;
            private Resources _resources;
            private URL _uri;
            private ServerFactory _factory;

            public MyListenerHandler( TcpTransportFactory ttf, URL uri, Resources resources, ServerFactory factory )
            {
                _ttf = ttf;
                _resources = resources;
                _uri = uri;
                _factory = factory;
            }

            #region ListenerHandler Members

            public void Accepted( Socket s )
            {
                Resources r = new Resources( _resources );
                r.Add( "SOCKET", s );
                r.Add( Messagizer.VALUE_FACTORY, _factory.NewValueFactory() );

                DefaultDeliveryService d = _ttf.NewTransport( _uri, r );

                _factory.NewServer( d );

                d.TransportControl( TransportConsts.START, null );
            }   

            #endregion

            #region SourceHandler<DataSource> Members

            private Listener _src;

            public Source GetSource()
            {
                return _src;
            }

            public void SetSource( Source src )
            {
                _src = (Listener) src;
            }

            #endregion

            #region Session Members

            public object SessionQuery( object query )
            {
                if ( _factory is Session )
                    return ( ( Session ) _factory ).SessionQuery( query );
                else
                    throw new Exception( "unknown query " + query );
            }

            public void SessionControl( object control, object value )
            {
                if ( _factory is Session )
                    ( ( Session ) _factory ).SessionControl( control, value );
                else
                    throw new Exception( "unknown control " + control );
            }

            public void SessionNotify( object eventObj )
            {
                if ( _factory is Session )
                    ( ( Session ) _factory ).SessionNotify( eventObj );
            }

            #endregion
        }

    }

    
}

File: src/etch/bindings/java/transport/TcpTransportFactory.java
package etch.bindings.java.transport;

import java.net.Socket;

import metreos.core.io.Listener;
import metreos.core.io.ListenerHandler;
import metreos.core.io.Packetizer;
import metreos.core.io.TcpConnection;
import metreos.util.URL;

/**
 * Transport factory for tcp.
 */
public class TcpTransportFactory extends TransportFactory
{
	public DefaultDeliveryService newTransport( URL uri,
		Resources resources ) throws Exception
	{
		Socket socket = (Socket) resources.get( "SOCKET" );
		
		TcpConnection c = new TcpConnection( uri, resources, socket );
		
		Packetizer p = new Packetizer( uri, resources );
		c.setHandler( p );
		
		Messagizer m = new Messagizer( uri, resources );
		p.setHandler( m );
		
		MailboxManager r = new MailboxManager( uri, resources );
		m.setHandler( r );
		
		DefaultDeliveryService d = new DefaultDeliveryService( uri, resources );
		r.setHandler( d );
		
		return d;
	}

	@Override
	public Transport newListener( final URL uri,
		final Resources resources, final ServerFactory factory )
		throws Exception
	{
		ListenerHandler lh = new ListenerHandler()
		{
			public void accepted( Socket s ) throws Exception
			{
				Resources r = new Resources( resources );
				r.put( "SOCKET", s );
				r.put( Messagizer.VALUE_FACTORY, factory.newValueFactory() );
				
				DefaultDeliveryService d = newTransport( uri, r );
				
				factory.newServer( d );
				
				d.transportControl( Transport.START, null );
			}

			public Listener getSource()
			{
				return src;
			}

			public void setSource( Listener src )
			{
				this.src = src;
			}
			
			private Listener src;

			public Object sessionQuery( Object query ) throws Exception
			{
				if (factory instanceof Session)
					return ((Session) factory).sessionQuery( query );
				else
					throw new UnsupportedOperationException( "unknown query "+query );
			}

			public void sessionControl( Object control, Object value ) throws Exception
			{
				if (factory instanceof Session)
					((Session) factory).sessionControl( control, value );
				else
					throw new UnsupportedOperationException( "unknown control "+control );
			}

			public void sessionNotify( Object event ) throws Exception
			{
				if (factory instanceof Session)
					((Session) factory).sessionQuery( event );
			}
		};

		Listener l = new Listener( uri );
		l.setHandler( lh );
		return l;
	}
}
